akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,140 @@
+package akka.http.impl.engine.client
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.Done
+import akka.actor.{ Actor, ActorRef, DeadLetterSuppression, Deploy, NoSerializationVerificationNeeded, Props, Terminated }
+import akka.http.impl.engine.client.PoolInterfaceActor.PoolRequest
+import akka.http.impl.settings.HostConnectionPoolSetup
+import akka.http.scaladsl.HttpExt
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
+import akka.stream.Materializer
+
+import scala.concurrent.Promise
+
+/**
+ * Manages access to a host connection pool or rather: a sequence of pool incarnations.
+ *
+ * A host connection pool for a given [[HostConnectionPoolSetup]] is a running stream, whose outside interface is
+ * provided by its [[PoolInterfaceActor]] actor. The actor accepts [[PoolInterfaceActor.PoolRequest]] messages
+ * and completes their `responsePromise` whenever the respective response has been received (or an error occurred).
+ *
+ * The [[PoolMasterActor]] provides a layer of indirection between a [[PoolGateway]], which represents a pool,
+ * and the [[PoolInterfaceActor]] instances which are created on-demand and stopped after an idle-timeout.
+ *
+ * Several [[PoolGateway]] objects may be mapped to the same pool if they have the same [[HostConnectionPoolSetup]]
+ * and are marked as being shared. This is the case for example for gateways obtained through
+ * [[HttpExt.cachedHostConnectionPool]]. Some other gateways are not shared, such as those obtained through
+ * [[HttpExt.newHostConnectionPool]], and will have their dedicated restartable pool.
+ *
+ */
+private[http] class PoolMasterActor extends Actor {
+
+  import PoolMasterActor._
+
+  private[this] implicit val dispatcher = context.dispatcher
```

Comment:
I'd avoid this, rather import `context.dispatcher` as close to where the ExecutionContext is needed as possible (to make it easier to reason about what parts uses it)